crates/matrix-sdk-common/src/timeout.rs

```diff
@@ -0,0 +1,51 @@
+use std::{error::Error, fmt, time::Duration};
+
+use futures_core::Future;
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::time::timeout;
+#[cfg(target_arch = "wasm32")]
+use wasm_timer::ext::TryFutureExt;
+
+/// Error type notifying that a timeout has elapsed.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ElapsedError(());
+
+impl fmt::Display for ElapsedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "time waiting for future has elapsed!")
+    }
+}
+
+impl Error for ElapsedError {}
+
+impl From<std::io::Error> for ElapsedError {
+    fn from(io_error: std::io::Error) -> Self {
+        io_error.into()
+    }
+}
+
+/// Wait for `future` to be completed. `future` needs to return
+/// a `Result`.
+///
+/// If the given timeout has elapsed the method will stop waiting and return
+/// an error.
+pub async fn wait<F, T>(future: F, duration: Duration) -> Result<T, ElapsedError>
```

Comment:
Why call this wait instead of `timeout` like the functions it wraps? I find it pretty confusing.

Reply:
I stuck to the original name of the function I moved out of `manager.rs`, I will change it to `timeout`.